File: frontend/src/mocks/factories/history.ts
/**
 * History factory - generates validation history and trends
 */

import type {
  HistoryResponse,
  HistorySummary,
  TrendDataPoint,
  FailureFrequencyItem,
  RecentValidation,
} from '@/api/client'
import { createId, randomChoice, randomInt, faker } from './base'

const ISSUE_TYPES = [
  'null_values',
  'type_mismatch',
  'out_of_range',
  'format_error',
  'duplicate_values',
  'missing_required',
  'constraint_violation',
]

function createSummary(totalRuns: number): HistorySummary {
  const passedRuns = Math.floor(totalRuns * faker.number.float({ min: 0.6, max: 0.95 }))
  const failedRuns = totalRuns - passedRuns

  return {
    total_runs: totalRuns,
    passed_runs: passedRuns,
    failed_runs: failedRuns,
    success_rate: (passedRuns / totalRuns) * 100,
  }
}

function createTrendData(
  period: '7d' | '30d' | '90d',
  granularity: 'hourly' | 'daily' | 'weekly'
): TrendDataPoint[] {
  const points: TrendDataPoint[] = []
  let days: number

  switch (period) {
    case '7d':
      days = 7
      break
    case '30d':
      days = 30
      break
    case '90d':
      days = 90
      break
  }

  const intervals = granularity === 'hourly' ? days * 24 : granularity === 'daily' ? days : Math.ceil(days / 7)

  for (let i = intervals - 1; i >= 0; i--) {
    const date = new Date()
    if (granularity === 'hourly') {
      date.setHours(date.getHours() - i)
    } else if (granularity === 'daily') {
      date.setDate(date.getDate() - i)
    } else {
      date.setDate(date.getDate() - i * 7)
    }

    const runCount = randomInt(5, 30)
    const passedCount = Math.floor(runCount * faker.number.float({ min: 0.6, max: 0.98 }))
    const failedCount = runCount - passedCount

    points.push({
      date: date.toISOString().split('T')[0],
      success_rate: (passedCount / runCount) * 100,
      run_count: runCount,
      passed_count: passedCount,
      failed_count: failedCount,
    })
  }

  return points
}

function createFailureFrequency(): FailureFrequencyItem[] {
  return faker.helpers
    .shuffle([...ISSUE_TYPES])
    .slice(0, randomInt(3, 6))
    .map((issue) => ({
      issue,
      count: randomInt(5, 100),
    }))
    .sort((a, b) => b.count - a.count)
}

function createRecentValidations(count: number): RecentValidation[] {
  const validations: RecentValidation[] = []

  for (let i = 0; i < count; i++) {
    const date = new Date()
    date.setHours(date.getHours() - i * randomInt(1, 6))

    const passed = faker.datatype.boolean(0.75)

    validations.push({
      id: createId(),
      status: passed ? 'success' : randomChoice(['failed', 'error']),
      passed,
      has_critical: !passed && faker.datatype.boolean(0.3),
      has_high: !passed && faker.datatype.boolean(0.5),
      total_issues: passed ? randomInt(0, 2) : randomInt(3, 20),
      created_at: date.toISOString(),
    })
  }

  return validations
}

export function createHistoryResponse(
  period: '7d' | '30d' | '90d' = '30d',
  granularity: 'hourly' | 'daily' | 'weekly' = 'daily'
): HistoryResponse {
  const totalRuns = randomInt(50, 500)

  return {
    success: true,
    data: {
      summary: createSummary(totalRuns),
      trend: createTrendData(period, granularity),
      failure_frequency: createFailureFrequency(),
      recent_validations: createRecentValidations(10),
    },
  }
}

File: src/truthound_dashboard/schemas/history.py
"""History and analytics schemas.

Schemas for validation history and trend analysis.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class TrendDataPoint(BaseModel):
    """Single data point in trend analysis."""

    date: str = Field(..., description="Date string (format depends on granularity)")
    success_rate: float = Field(..., ge=0, le=100, description="Success rate percentage")
    run_count: int = Field(..., ge=0, description="Number of validation runs")
    passed_count: int = Field(..., ge=0, description="Number of passed validations")
    failed_count: int = Field(..., ge=0, description="Number of failed validations")


class FailureFrequencyItem(BaseModel):
    """Failure frequency for an issue type."""

    issue: str = Field(..., description="Issue identifier (column.issue_type)")
    count: int = Field(..., ge=0, description="Total occurrence count")


class RecentValidation(BaseModel):
    """Summary of a recent validation run."""

    id: str = Field(..., description="Validation ID")
    status: str = Field(..., description="Validation status")
    passed: bool | None = Field(None, description="Whether validation passed")
    has_critical: bool | None = Field(None, description="Has critical issues")
    has_high: bool | None = Field(None, description="Has high severity issues")
    total_issues: int | None = Field(None, description="Total issue count")
    created_at: str = Field(..., description="ISO timestamp")


class HistorySummary(BaseModel):
    """Summary statistics for the history period."""

    total_runs: int = Field(..., ge=0, description="Total validation runs")
    passed_runs: int = Field(..., ge=0, description="Number of passed runs")
    failed_runs: int = Field(..., ge=0, description="Number of failed runs")
    success_rate: float = Field(..., ge=0, le=100, description="Success rate percentage")


class HistoryResponse(BaseModel):
    """Validation history response with trends and analytics."""

    summary: HistorySummary = Field(..., description="Summary statistics")
    trend: list[TrendDataPoint] = Field(default_factory=list, description="Trend data")
    failure_frequency: list[FailureFrequencyItem] = Field(
        default_factory=list, description="Top failure types"
    )
    recent_validations: list[RecentValidation] = Field(
        default_factory=list, description="Recent validation runs"
    )


class HistoryQueryParams(BaseModel):
    """Query parameters for history endpoint."""

    period: Literal["7d", "30d", "90d"] = Field("30d", description="Time period")
    granularity: Literal["hourly", "daily", "weekly"] = Field(
        "daily", description="Aggregation granularity"
    )
